library/core/src/hash/mod.rs

```diff
@@ -960,17 +960,21 @@ mod impls {
         fn hash<H: Hasher>(&self, state: &mut H) {
             let (address, metadata) = self.to_raw_parts();
             state.write_usize(address.addr());
-            metadata.hash(state);
+            if mem::size_of_val(&metadata) == mem::size_of::<usize>() {
```

Comment:
surely it would be better to just require metadata to be `Hash` instead of special casing `usize` size'd metadata via a transmute. this also means that we cannot have `usize` sized metadata with padding it

Reply:
doing this would make it extremely hard to ever support unsizing away arbitrary const generics with `adt_const_params` since if you had a `usize` size'd type for your const arg it would now implicitly get hashed whereas if you added a byte of padding you wouldnt.

Reply:
Perhaps we could have an intrinsic like `type_id` that works for non 'static types that just returns `bool` for whether the metadata here is a `usize` instead of assuming that the only `usize` size'd metadata we are ever going to have is a `usize`

Reply:
I'd be in favor of that, the impl is kind of gross as is.

Reply:
I guess the metadata can also be pointers to vtables which would be `usize` sized, is this supposed to affect those too? I guess the intrinsic would have to be more complex if so

Reply:
It'll be confusing to have `DynMetadata` of the same type to hash to different values.

The impl here is essentially just going back to it's old state before we have the unstable ptr metadata API (the old one uses ptr casting while I chose to use transmute_copy, but the essence is the same).

One possible solution to avoid this is to have a private-to-core trait `MetadataHash` that's implemented both for `()`, `usize` and `DynMetadata`, and we have that in the bound instead of `Hash`.

Reply:
@nbdd0121 just so I'm clear: What is the longer-term goal w.r.t. to `*dyn Trait` metadata (i.e. pointers to vtables)?

My understanding is that this PR continues to hash them (by hashing their raw address, transmuted to a `usize`), but it may be safer to exclude them, due to the issues pointed out in https://github.com/rust-lang/rust/issues/106447#issuecomment-1566147477 .

Excluding the metadata in that case will make hashes collide for two distinct `&dyn Trait1` and `&dyn Trait2` that has the same pointer address, but... I think that's acceptable?

Reply:
> It'll be confusing to have `DynMetadata` of the same type to hash to different values.

Also, I didn't understand what the "of the same type" was saying here. The same Type for distinct Trait1/Trait2? Or the same instance of the vtable (i.e. the same address)? Or the same Trait1 but distinct vtables (and thus the distinct hash values)?